### ISSUE TITLE
942250 (SQLi HAVING.. rule) has lots of false positives

### ISSUE THREAD

**Issue by lifeforms** [SpiderLabs/owasp-modsecurity-crs#409]
Rule 942250 (Detects MATCH AGAINST, MERGE, EXECUTE IMMEDIATE and HAVING injections) has 2% false positives on the Reddit comments dataset (See #318 for links and discussion)

Can we lower this?

Analyze the matched data on Reddit comments.

**Comment by dune73** [comment]
I think HAVING by itself is harmless:

```
$> grep -c curl sqlmap-attack-traffic.sh 
20781
$> grep -i having sqlmap-attack-traffic.sh  | wc -l
99
$> grep -i having sqlmap-attack-traffic.sh  | grep -v -i select | wc -l
0
```

Example payloads:
```
 -3223%'))) OR 1 GROUP BY CONCAT(0x716a787a71,(SELECT (CASE WHEN (3900=3900) THEN 1 ELSE 0 END)),0x7171627671,FLOOR(RAND(0)*2)) HAVING MIN(0)#" -H 'Accept
 -2906%' OR 1 GROUP BY CONCAT(0x716a787a71,(SELECT (CASE WHEN (3900=3900) THEN 1 ELSE 0 END)),0x7171627671,FLOOR(RAND(0)*2)) HAVING MIN(0)#" -H 'Accept
 -2836%00' OR 1 GROUP BY CONCAT(0x716a787a71,(SELECT (CASE WHEN (3900=3900) THEN 1 ELSE 0 END)),0x7171627671,FLOOR(RAND(0)*2)) HAVING MIN(0)#" -H 'Accept
 -8115 OR 1 GROUP BY CONCAT(0x716a787a71,(SELECT (CASE WHEN (3900=3900) THEN 1 ELSE 0 END)),0x7171627671,FLOOR(RAND(0)*2)) HAVING MIN(0)#" -H 'Accept
 -2949 OR 1 GROUP BY CONCAT(0x716a787a71,(SELECT (CASE WHEN (3900=3900) THEN 1 ELSE 0 END)),0x7171627671,FLOOR(RAND(0)*2)) HAVING MIN(0)#" -H 'Accept
 -4609) WHERE 5091=5091 OR 1 GROUP BY CONCAT(0x716a787a71,(SELECT (CASE WHEN (3900=3900) THEN 1 ELSE 0 END)),0x7171627671,FLOOR(RAND(0)*2)) HAVING MIN(0)#" -H 

```

**Comment by dune73** [comment]
Not only SELECT, but also GROUP BY are preconditions for the use of HAVING in an sql statement.

http://www.w3schools.com/sql/sql_having.asp

**Comment by lifeforms** [comment]
In MySQL, you can just tack `HAVING` onto some query without groups. For instance let's have this query: `select * from foo where name=$name`. If you would pass `name=name having 1` it behaves the same as if you would have said `name=name or 1`.

Now I'm not an absolute SQLi expert. The big question is, whether rule 942250 is essential seeing that libinjection provides base coverage. It seems that libinjection (rule 942100) does look at `HAVING` constructs. In PL1 I see the following:

```
sco rules                               payload
--- ----------------------------------- -----------------------------------
0                                       1 or 1
5   942100                              1 or 1=1
10  942100,942250                       1 having 1
15  942100,942230,942250                1 having 1=1
```

(By the way, pretty interesting that libinjection doesn't think the first test is an injection, I think we should detect that form in PL1, but that's another story.)

I also checked with sqlmap. It has `HAVING` in its toolbox. I did many scans with various settings. I checked in the audit log whenever rule 942250 (HAVING...) was being hit. In all of these cases, the total score was 20. This would suggest that even when removing HAVING from the regexp, the payloads should be caught anyway.

Seeing that it's the rule causing most FP on the Reddit dataset at the moment, and it seems non-essential, I agree with @dune73 that we should move this to PL2.

It will require a bit of manual labor to extract the HAVING pattern from the other patterns in the regexp, so please ack before I open a PR.

**Comment by lifeforms** [comment]
Disregard that, maybe even move the pattern to PL3.

* It seems non-essential
* It blocks every parameter where people use the word `having` in a sentence.

**Comment by dune73** [comment]
Seems warranted. Thanks for looking at this in detail.

**Comment by lifeforms** [comment]
Fixed in #441
